File: data_structure/python/graph.py
#---------------------------------------------------------------
# GRAPH (adjacency list)
#---------------------------------------------------------------
#
# A graph is a set of vertices plus a set of edges. Unlike a tree it
# may contain cycles and has no root.
#
#         0 --- 1 --- 3
#         |    /|     |
#         |   / |     |
#         2 -/  |     4 --- 5
#                \         /
#                 --------- 6
#
# TWO WAYS TO STORE IT
#
#   adjacency LIST    {0: [1, 2], 1: [0, 2, 3], ...}
#     space O(V + E), listing a vertex's neighbours is O(deg(v))
#     -> the default for SPARSE graphs, which is nearly every
#        interview/LeetCode graph
#
#   adjacency MATRIX  matrix[i][j] == 1 when edge i-j exists
#     space O(V^2), edge lookup is O(1), neighbour listing is O(V)
#     -> only worth it for DENSE graphs or O(1) edge tests
#
# Time  : add_vertex / add_edge  O(1)
#         has_edge               O(deg(v))
#         bfs / dfs              O(V + E)
# Space : O(V + E)
#
# References:
#   - https://github.com/yennanliu/CS_basics/blob/master/data_structure/js/graph.js


from collections import deque


class Graph:
    """Undirected graph stored as an adjacency list."""

    def __init__(self, directed=False):
        self.directed = directed
        self.adjacent = {}

    def __str__(self):
        return "\n".join(
            "{} --> {}".format(node, " ".join(str(n) for n in sorted(neighbours)))
            for node, neighbours in sorted(self.adjacent.items(), key=lambda kv: str(kv[0]))
        )

    @property
    def vertex_count(self):
        return len(self.adjacent)

    @property
    def edge_count(self):
        total = sum(len(n) for n in self.adjacent.values())
        return total if self.directed else total // 2

    def add_vertex(self, node):
        """Adding an isolated vertex is NOT the same as adding an edge."""
        self.adjacent.setdefault(node, [])

    def add_edge(self, node1, node2):
        """Connect two vertices, creating them if they do not exist yet."""
        self.add_vertex(node1)
        self.add_vertex(node2)
        self.adjacent[node1].append(node2)
        if not self.directed:            # undirected -> record BOTH directions
            self.adjacent[node2].append(node1)

    def neighbours(self, node):
        return self.adjacent.get(node, [])

    def has_edge(self, node1, node2):
        return node2 in self.adjacent.get(node1, [])

    def bfs(self, start):
        """Breadth-first: nearest vertices first, using a QUEUE."""
        if start not in self.adjacent:
            return []
        visited, order, queue = {start}, [], deque([start])
        while queue:
            node = queue.popleft()
            order.append(node)
            for neighbour in self.adjacent[node]:
                if neighbour not in visited:
                    visited.add(neighbour)   # mark on ENQUEUE, or a vertex
                    queue.append(neighbour)  # can be queued twice
        return order

    def dfs(self, start):
        """Depth-first: follow one branch to the end, using recursion."""
        order, visited = [], set()

        def walk(node):
            visited.add(node)
            order.append(node)
            for neighbour in self.adjacent.get(node, []):
                if neighbour not in visited:
                    walk(neighbour)

        if start in self.adjacent:
            walk(start)
        return order

    def shortest_path(self, start, end):
        """Fewest EDGES from start to end (BFS), or None if unreachable.

        Only correct on an UNWEIGHTED graph -- with weights you need
        Dijkstra (see algorithm/python/dijkstra.py).
        """
        if start not in self.adjacent or end not in self.adjacent:
            return None
        if start == end:
            return [start]
        previous, queue = {start: None}, deque([start])
        while queue:
            node = queue.popleft()
            for neighbour in self.adjacent[node]:
                if neighbour in previous:
                    continue
                previous[neighbour] = node
                if neighbour == end:                 # rebuild the path backwards
                    path = [end]
                    while previous[path[-1]] is not None:
                        path.append(previous[path[-1]])
                    return path[::-1]
                queue.append(neighbour)
        return None


if __name__ == "__main__":
    #    0 --- 1 --- 3
    #    |    /|     |
    #    2 --/ |     4 --- 5 --- 6
    g = Graph()
    for edge in [("0", "1"), ("0", "2"), ("1", "2"), ("1", "3"),
                 ("3", "4"), ("4", "5"), ("5", "6")]:
        g.add_edge(*edge)

    assert g.vertex_count == 7
    assert g.edge_count == 7
    assert sorted(g.neighbours("1")) == ["0", "2", "3"]
    assert g.has_edge("0", "1") and g.has_edge("1", "0")   # undirected
    assert not g.has_edge("0", "6")

    assert g.bfs("0") == ["0", "1", "2", "3", "4", "5", "6"]
    assert g.dfs("0") == ["0", "1", "2", "3", "4", "5", "6"]

    assert g.shortest_path("0", "6") == ["0", "1", "3", "4", "5", "6"]
    assert g.shortest_path("0", "0") == ["0"]
    g.add_vertex("island")
    assert g.shortest_path("0", "island") is None

    # in a DIRECTED graph the edge goes one way only
    d = Graph(directed=True)
    d.add_edge("a", "b")
    assert d.has_edge("a", "b") and not d.has_edge("b", "a")
    assert d.edge_count == 1

    print(g)
    print("Success.")

File: data_structure/js/graph.js
//---------------------------------------------------------------
// GRAPH (adjacency list)
//---------------------------------------------------------------
//
// A graph is a set of vertices plus a set of edges. Unlike a tree it
// may contain cycles and has no root.
//
//     0 --- 1 --- 3
//     |    /|     |
//     2 --/ |     4 --- 5 --- 6
//
// TWO WAYS TO STORE IT
//
//   adjacency LIST    { 0: [1, 2], 1: [0, 2, 3], ... }
//     space O(V + E); listing a vertex's neighbours is O(deg(v))
//     -> the default for SPARSE graphs, which is nearly every
//        interview graph
//
//   adjacency MATRIX  matrix[i][j] === 1 when the edge i-j exists
//     space O(V^2); edge lookup is O(1), neighbour listing is O(V)
//     -> only worth it for DENSE graphs or O(1) edge tests
//
// Time  : addVertex / addEdge  O(1)
//         hasEdge              O(deg(v))
//         bfs / dfs            O(V + E)
// Space : O(V + E)

class Graph {
  constructor(directed = false) {
    this.directed = directed;
    this.adjacentList = {};
  }

  get numberOfNodes() {
    return Object.keys(this.adjacentList).length;
  }

  get numberOfEdges() {
    const total = Object.values(this.adjacentList).reduce((n, list) => n + list.length, 0);
    return this.directed ? total : total / 2;
  }

  // An isolated vertex is NOT the same as an edge.
  addVertex(node) {
    if (!this.adjacentList[node]) this.adjacentList[node] = [];
    return this;
  }

  addEdge(node1, node2) {
    this.addVertex(node1);
    this.addVertex(node2);
    this.adjacentList[node1].push(node2);
    if (!this.directed) this.adjacentList[node2].push(node1); // record BOTH directions
    return this;
  }

  neighbours(node) {
    return this.adjacentList[node] || [];
  }

  hasEdge(node1, node2) {
    return this.neighbours(node1).includes(node2);
  }

  // Breadth-first: nearest vertices first, using a QUEUE.
  bfs(start) {
    if (!this.adjacentList[start]) return [];
    const visited = new Set([start]);
    const order = [];
    const queue = [start];

    while (queue.length > 0) {
      const node = queue.shift();
      order.push(node);
      for (const neighbour of this.adjacentList[node]) {
        if (visited.has(neighbour)) continue;
        visited.add(neighbour); // mark on ENQUEUE, or a vertex can be queued twice
        queue.push(neighbour);
      }
    }
    return order;
  }

  // Depth-first: follow one branch to the end, using recursion.
  dfs(start) {
    const visited = new Set();
    const order = [];

    const walk = (node) => {
      visited.add(node);
      order.push(node);
      for (const neighbour of this.neighbours(node)) {
        if (!visited.has(neighbour)) walk(neighbour);
      }
    };

    if (this.adjacentList[start]) walk(start);
    return order;
  }

  // Fewest EDGES from start to end (BFS), or null if unreachable.
  // Only correct on an UNWEIGHTED graph -- with weights you need Dijkstra.
  shortestPath(start, end) {
    if (!this.adjacentList[start] || !this.adjacentList[end]) return null;
    if (start === end) return [start];

    const previous = new Map([[start, null]]);
    const queue = [start];

    while (queue.length > 0) {
      const node = queue.shift();
      for (const neighbour of this.adjacentList[node]) {
        if (previous.has(neighbour)) continue;
        previous.set(neighbour, node);
        if (neighbour === end) {
          const path = [end]; // rebuild the path backwards
          while (previous.get(path[path.length - 1]) !== null) {
            path.push(previous.get(path[path.length - 1]));
          }
          return path.reverse();
        }
        queue.push(neighbour);
      }
    }
    return null;
  }

  showConnections() {
    return Object.keys(this.adjacentList)
      .sort()
      .map((node) => `${node}-->${this.adjacentList[node].join(' ')}`)
      .join('\n');
  }
}

// demo
//    0 --- 1 --- 3
//    |    /|     |
//    2 --/ |     4 --- 5 --- 6
const graph = new Graph();
[['3', '1'], ['3', '4'], ['4', '2'], ['4', '5'],
 ['1', '2'], ['1', '0'], ['0', '2'], ['6', '5']].forEach(([a, b]) => graph.addEdge(a, b));

console.assert(graph.numberOfNodes === 7, 'seven vertices');
console.assert(graph.numberOfEdges === 8, 'eight edges');
console.assert(graph.hasEdge('3', '1') && graph.hasEdge('1', '3'), 'undirected: both ways');
console.assert(!graph.hasEdge('0', '6'), 'no direct edge');
// copy before sorting -- .sort() mutates, and neighbours() returns the live array
console.assert([...graph.neighbours('1')].sort().join(',') === '0,2,3');

console.assert(graph.bfs('0').join(',') === '0,1,2,3,4,5,6', 'BFS from 0');
console.assert(graph.dfs('0').join(',') === '0,1,3,4,2,5,6', 'DFS from 0');

console.assert(graph.shortestPath('0', '6').join(',') === '0,2,4,5,6', 'fewest edges');
console.assert(graph.shortestPath('0', '0').join(',') === '0', 'path to self');
graph.addVertex('island');
console.assert(graph.shortestPath('0', 'island') === null, 'unreachable -> null');

// a DIRECTED edge goes one way only
const digraph = new Graph(true);
digraph.addEdge('a', 'b');
console.assert(digraph.hasEdge('a', 'b') && !digraph.hasEdge('b', 'a'), 'directed');
console.assert(digraph.numberOfEdges === 1);

console.log(graph.showConnections());
console.log('Success.');

module.exports = { Graph };
